src/plot_api/plot_api.js

```diff
@@ -1772,6 +1683,26 @@ exports.relayout = function relayout(gd, astr, val) {
 
         if(flags.legend) seq.push(subroutines.doLegend);
         if(flags.layoutstyle) seq.push(subroutines.layoutStyles);
+
+        if(flags.axrange) {
+            // N.B. leave as sequence of subroutines (for now) instead of
+            // subroutine of its own so that finalDraw always gets
+            // executed after drawData
+            seq.push(
+                // TODO
+                // no test fail when commenting out doAutoRangeAndConstraints,
+                // but I think we do need this (maybe just the enforce part?)
+                // Am I right?
```

Comment:
Huh, I'd have thought so... perhaps worth spending a little time looking for something that does fail if this is removed, so it can be 🔒 down.

Reply:
@alexcjohnson Tests are currently passing even when `subroutines.doAutoRangeAndConstraints` is commented out, because of this block:

https://github.com/plotly/plotly.js/blob/2a2582022105abd640ef3e5a38d98fb343a06779/src/plot_api/plot_api.js#L2088-L2107

which makes axis range relayout calls altering constrained axes go though the `calc: true` edit pathway which (of course) calls `doAutoRangeAndConstraints` itself.

-------------

Interestingly, commenting out that [block](https://github.com/plotly/plotly.js/blob/2a2582022105abd640ef3e5a38d98fb343a06779/src/plot_api/plot_api.js#L2088-L2107) which sets `flags.calc = true` does not make any test fails when `subroutines.doAutoRangeAndConstraints` is called during `axrange` edits. Perhaps we don't need?

Reply:
It doesn't look to me as though I wrote all that great test coverage for this section... the two key points of it are:
- You can `relayout` a constrained axis, particularly to a smaller range, and have the other axes in the group update (expanding *or* shrinking) to continue to meet the constraints (note that this won't work with `react`, there you'd have to manually update *all* axes in the constraint group)
- If you `relayout` (or `react`) two or more axes in a group in such a way as to explicitly violate the constraints, one or more of *those* axes will expand (never shrink) its range to fit the constraints.

So if you want to clean this up we should first add a few tests explicitly for ^^. I see some [gui tests](https://github.com/plotly/plotly.js/pull/1522/files#diff-200dbdf53958cf8973413a709518f4bbR389) and one [restyle test](https://github.com/plotly/plotly.js/pull/2465/files#diff-5a140ce1cc87a420496039bbf8d699f1R1265) that uses constraints, but unfortunately no `relayout` or `react` tests of these edge cases.

Reply:
Ok if I open up an issue about this and defer this to a later PR?

Reply:
> Ok if I open up an issue about this and defer this to a later PR?

Yep for sure - that was the "if you want to clean this up" clause

Reply:
Moved to -> https://github.com/plotly/plotly.js/issues/2540